### PR TITLE
Delay drag start until threshold, add force simulation on drag

### PR DESCRIPTION
**Pull request by jeff-phillips-18**

### CODE REVIEW

File: frontend/packages/topology/src/behavior/useDndDrag.tsx [hunk 1]
```diff
@@ -74,65 +81,79 @@ export const useDndDrag = <
...
+                action(() => {
+                  if (
+                    dndManagerRef.current.drag(
+                      d3.event.x,
+                      d3.event.y,
+                      d3.event.sourceEvent.pageX,
+                      d3.event.sourceEvent.pageY,
+                    )
+                  ) {
+                    controller.fireEvent(DRAG_EVENT, id, d3.event.x, d3.event.y);
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/useDndDrag.tsx]: I think the events should fire with the `DndState` object. This is currently private inside the DndManager but i'm fine exposing it as `getState()`.

Note that DnD is currently entity agnostic in that the payload of drag doesn't necessarily include the entity. It's up to the setup of the drag to include a payload that a drop zone will understand.
Looks like you need the entity ID to understand if it is being dragged. Will need to brainstorm here a bit to think of a good solution.

File: frontend/packages/topology/src/behavior/useDndDrag.tsx [hunk 2]
```diff
@@ -74,65 +81,79 @@ export const useDndDrag = <
...
+                      d3.event.y,
+                      d3.event.sourceEvent.pageX,
+                      d3.event.sourceEvent.pageY,
+                      () => {
+                        controller.fireEvent(DRAG_START_EVENT, id);
+                      },
+                    );
+                  }
+                }),
+              )
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/useDndDrag.tsx]: We can delete this if we make the d3 `drag` event handle call `beginDrag`

File: frontend/packages/topology/src/behavior/useDndDrag.tsx [hunk 3]
```diff
@@ -74,65 +81,79 @@ export const useDndDrag = <
...
+                'drag',
+                action(() => {
+                  if (
+                    dndManagerRef.current.drag(
+                      d3.event.x,
+                      d3.event.y,
+                      d3.event.sourceEvent.pageX,
+                      d3.event.sourceEvent.pageY,
+                    )
+                  ) {
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/useDndDrag.tsx]: Why not simplify the entire process to:
```
if (dndManager.isDragging()) {
  dndManagerRef.current.drag(...);
  entityRef.current.getController().fireEvent(DRAG_EVENT, dndManagerRef.current.getState());
} else if (d3.event.dx > 5 || d3.event.dy > 5) {
  dndManagerRef.current.beginDrag(...);
  entityRef.current.getController().fireEvent(DRAG_START_EVENT, dndManagerRef.current.getState());
}
```

File: frontend/packages/topology/src/behavior/dnd-types.ts [hunk 4]
```diff
@@ -42,15 +42,34 @@ export type DndStateContainer = {
...
 export type Unregister = () => void;
-export interface DndManager extends DndActions {
+export interface DndManager {
+  beginDrag(
+    sourceIds: string | string[],
+    x: number,
+    y: number,
+    pageX: number,
+    pageY: number,
+    startCallback?: () => void,
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/dnd-types.ts]: You shouldn't need this if you look at my other comments.

**Comment by jeff-phillips-18** [comment]
Simplified a lot. Moved the event firing to `useDrageNode`, got rid of extra vars in `useDndManager`.

**Pull request merged**